### PR TITLE
add auditing

### PR DESCRIPTION
**Pull request by wanjunlei**
What type of PR is this?
/kind feature

What this PR does / why we need it:
This PR will add auditing in ks-apiserver. Each request will generate an auditing event, and the event will be sent to kube-auditing-webhook by ks-apiserver.  The auditing rule determines which event will be persisted and which event will triggering an alert.

### CODE REVIEW

File: pkg/simple/client/auditing/elasticsearch/options.go [hunk 1]
```diff
@@ -47,6 +48,7 @@ func (s *Options) Validate() []error {
 }
 
 func (s *Options) AddFlags(fs *pflag.FlagSet, c *Options) {
+	fs.BoolVar(&s.Enabled, "auditing-enabled", c.Enabled, "Dose auditing enabled")
```
Comment (zryfish) [inline pkg/simple/client/auditing/elasticsearch/options.go:51]: Typo `Does`,  I think we can following description.
```suggestion
	fs.BoolVar(&s.Enabled, "auditing-enabled", c.Enabled, "Enable auditing component or not.")
```

File: pkg/simple/client/auditing/elasticsearch/options.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ import (
 )
 
 type Options struct {
+	Enabled     bool   `json:"enabled" yaml:"enable"`
```
Comment (zryfish) [inline pkg/simple/client/auditing/elasticsearch/options.go:25]: Why json tag is `enabled` while yam tag is `enable`, this is misleading. I think `enable` will be just fine.

File: pkg/simple/client/auditing/elasticsearch/options.go [hunk 3]
```diff
@@ -47,6 +48,7 @@ func (s *Options) Validate() []error {
 }
 
 func (s *Options) AddFlags(fs *pflag.FlagSet, c *Options) {
+	fs.BoolVar(&s.Enabled, "auditing-enabled", c.Enabled, "Dose auditing enabled")
 	fs.StringVar(&s.Host, "auditing-elasticsearch-host", c.Host, ""+
 		"Elasticsearch service host. KubeSphere is using elastic as auditing store, "+
 		"if this filed left blank, KubeSphere will use kubernetes builtin event API instead, and"+
```
Comment (zryfish) [inline pkg/simple/client/auditing/elasticsearch/options.go:54]: Have we already implemented this, use built-in event API when our auditing component not enabled?

File: pkg/apiserver/filters/auditing.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	"kubesphere.io/kubesphere/pkg/apiserver/auditing"
+	"kubesphere.io/kubesphere/pkg/apiserver/request"
+	"net/http"
+)
+
+func WithAuditing(handler http.Handler, a auditing.Auditing) http.Handler {
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+
+		if !a.Enable() {
```
Comment (zryfish) [inline pkg/apiserver/filters/auditing.go:14]: it's unnecessary to do a check here, just return `handler` when auditing is not enabled. Check out the code above

File: pkg/apiserver/filters/auditing.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package filters
+
+import (
+	"k8s.io/klog"
+	"kubesphere.io/kubesphere/pkg/apiserver/auditing"
+	"kubesphere.io/kubesphere/pkg/apiserver/request"
+	"net/http"
+)
+
+func WithAuditing(handler http.Handler, a auditing.Auditing) http.Handler {
```
Comment (zryfish) [inline pkg/apiserver/filters/auditing.go:10]:
```suggestion
func WithAuditing(handler http.Handler, a auditing.Auditing) http.Handler {
   if !a.Enable() {
      return handler
   }
}
```

File: pkg/apiserver/filters/auditing.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+		if !a.Enable() {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		info, ok := request.RequestInfoFrom(req.Context())
+		if !ok {
+			klog.Error("Unable to retrieve request info from request")
+			handler.ServeHTTP(w, req)
```
Comment (zryfish) [inline pkg/apiserver/filters/auditing.go:22]: If we can't retrieve request info from context, why just let the request pass, shouldn't we return error ?

File: pkg/apiserver/filters/auditing.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+		}
+
+		if info.IsKubernetesRequest {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		e := a.LogRequestObject(info, req)
+		resp := auditing.NewResponseCapture(w)
+		go handler.ServeHTTP(resp, req)
```
Comment (zryfish) [inline pkg/apiserver/filters/auditing.go:33]: Why put handler.ServeHTTP in a goroutine, why we need to wait from resp.StopCh ?
You can refer to https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/filters/audit.go#L77 to make sure `audit LogResponseObject` is always called after finish serving requests?

File: pkg/apiserver/auditing/backend.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	go b.worker()
+
+	return &b
+}
+
+func (b *Backend) worker() {
+
+	for {
```
Comment (zryfish) [inline pkg/apiserver/auditing/backend.go:56]: how do we end this loop? There is only one condition to end this loop `event == nil`, but when program is gracefully terminated, how can we end this loop gracefully?

File: pkg/apiserver/auditing/config.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"k8s.io/klog"
+	"os"
+	"regexp"
+)
+
+const (
+	ConfigFile = "/etc/kubesphere/auditing/config"
```
Comment (zryfish) [inline pkg/apiserver/auditing/config.go:12]: Is there a default auditing config file?

**Comment by wanjunlei** [comment]
There has a big change, i will resubumit this pr.

**Pull request closed without merging**